File: Terminal/TerminalLib/ServerExchange/PacketToRawData.cpp
#include "stdafx.h"
#include "PacketToRawData.h"

using namespace server_exchange;

CPacketToRawData::CPacketToRawData()
{
}


CPacketToRawData::~CPacketToRawData()
{
}

e_convert_result CPacketToRawData::CreateRawData(IN const tag_transport_packet& packet, 
												 OUT tools::data_wrappers::_tag_data_managed& result_data)
{
	uint32_t packet_size = 9 + packet.length;

	result_data.alloc_data(packet_size);

	byte* p_data = result_data.p_data;

	*(uint16_t*)p_data = packet.begin;
	p_data += 2;

	*p_data = static_cast<byte>(packet.type);
	++p_data;

	*(uint32_t*)p_data = packet.length;
	p_data += 4;

	if (0 != memcpy_s(p_data, packet_size - 9, packet.data.p_data, packet.data.data_size))
	{
		return e_convert_result::invalid_data;
	}

	p_data += packet.data.data_size;

	*(uint16_t*)p_data = packet.end;

	return e_convert_result::success;
}

e_convert_result CPacketToRawData::CreateLogRecordPacketRawData(IN const tag_log_record_packet& packet, 
																OUT tools::data_wrappers::_tag_data_managed& result_data)
{
	uint32_t packet_size = (11 + packet.text.data_size);

	result_data.alloc_data(packet_size);

	byte* p_data = result_data.p_data;

	*(uint64_t*)p_data = packet.date_time;
	p_data += sizeof(packet.date_time);

	*p_data = static_cast<byte>(packet.type);
	p_data += sizeof(packet.type);

	*(uint16_t*)p_data = packet.length;
	p_data += sizeof(packet.length);

	if (0 != memcpy_s(p_data, packet_size - 11, packet.text.p_data, packet.text.data_size))
	{
		return e_convert_result::invalid_data;
	}

	return e_convert_result::success;
}

File: Parsing/PacketToRawData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Parsing
{
    public class PacketToRawData
    {

        // создание данных для отправки из пакета транспортного уровня
	    public e_convert_result CreateRawData(tag_transport_packet packet, out Byte[] result_data)
        {
            List<Byte> collection = new List<Byte>();

            collection.AddRange(BitConverter.GetBytes(packet.begin));
            collection.Add((Byte)packet.type);
            collection.AddRange(BitConverter.GetBytes(packet.length));
            collection.AddRange(packet.data);
            collection.AddRange(BitConverter.GetBytes(packet.end));

            result_data = collection.ToArray();

            return e_convert_result.success;
        }

        // создание данных для отправки из пакета идентификации
        public e_convert_result CreateIdentificationPacketRawData(tag_identification_packet packet, out Byte[] result_data)
        {
            List<Byte> collection = new List<Byte>();

            collection.AddRange(packet.group_name.ToByteArray());
            collection.AddRange(packet.terminal_name.ToByteArray());

            result_data = collection.ToArray();
            return e_convert_result.success;
        }

        // создание данных для отправки из пакета счётчиков
        public e_convert_result CreateCountersPacketRawData(tag_counters_packet packet, out Byte[] result_data)
        {
            List<Byte> collection = new List<Byte>();

            collection.AddRange(BitConverter.GetBytes(packet.date_time));
            collection.AddRange(BitConverter.GetBytes(packet.total_cache));
            collection.AddRange(BitConverter.GetBytes(packet.current_cache));
            collection.Add((Byte)packet.state);
            collection.AddRange(BitConverter.GetBytes(packet.pressurized_water));
            collection.AddRange(BitConverter.GetBytes(packet.water_without_pressure));
            collection.AddRange(BitConverter.GetBytes(packet.foam));
            collection.AddRange(BitConverter.GetBytes(packet.wax));
            collection.AddRange(BitConverter.GetBytes(packet.against_midges));
            collection.AddRange(BitConverter.GetBytes(packet.vacuum_cleaner));
            collection.AddRange(BitConverter.GetBytes(packet.air));
            collection.AddRange(BitConverter.GetBytes(packet.osmosis));

            result_data = collection.ToArray();
            return e_convert_result.success;
        }

        // создание данных для отправки из пакета настроек
        public e_convert_result CreateSettingsPacketRawData(tag_settings_packet packet, out Byte[] result_data)
        {
            List<Byte> collection = new List<Byte>();

            collection.Add((Byte)packet.state);
            collection.Add(packet.bill_acceptor_impulse);
            collection.Add(packet.coin_acceptor_impulse);
            collection.Add(packet.free_idle_time);
            collection.AddRange(BitConverter.GetBytes(packet.idle_time_cost));
            collection.Add(packet.pause_before_advertising);
            collection.AddRange(BitConverter.GetBytes(packet.pressurized_water));
            collection.AddRange(BitConverter.GetBytes(packet.water_without_pressure));
            collection.AddRange(BitConverter.GetBytes(packet.foam));
            collection.AddRange(BitConverter.GetBytes(packet.wax));
            collection.AddRange(BitConverter.GetBytes(packet.against_midges));
            collection.AddRange(BitConverter.GetBytes(packet.vacuum_cleaner));
            collection.AddRange(BitConverter.GetBytes(packet.air));
            collection.AddRange(BitConverter.GetBytes(packet.osmosis));

            result_data = collection.ToArray();
            return e_convert_result.success;
        }

        // создание данных для отправки из пакета настроек
        public e_convert_result CreateConfirmationPacketRawData(tag_confirmation_packet packet, out Byte[] result_data)
        {
            result_data = new Byte[2];

            result_data[0] = (Byte)packet.packet_type;
            result_data[1] = (Byte)packet.result;

            return e_convert_result.success;
        }

        // создание данных для отправки из пакета пополнения баланса
        public e_convert_result CreateRefillCachePacketRawData(tag_refill_cache_packet packet, out Byte[] result_data)
        {
            result_data = BitConverter.GetBytes(packet.cache);

            return e_convert_result.success;
        }

        // создание данных для отправки из пакета приза
        public e_convert_result CreatePrizePacketRawData(tag_prize_packet packet, out Byte[] result_data)
        {
            result_data = BitConverter.GetBytes(packet.prize_size);

            return e_convert_result.success;
        }

        // создание данных для отправки из пакета настроек
        public e_convert_result CreateLogRecordPacketRawData(tag_log_record_packet packet, out Byte[] result_data)
        {
            List<Byte> collection = new List<Byte>();

            collection.AddRange(BitConverter.GetBytes(packet.date_time));
            collection.Add((Byte)packet.type);
            collection.AddRange(BitConverter.GetBytes(packet.length));

            foreach (Char c in packet.text)
            {
                collection.AddRange(BitConverter.GetBytes(c));
            }

            result_data = collection.ToArray();

            return e_convert_result.success;
        }


    }
}
